clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
Nice catch! I think you are referring to `fetcher.initFetches(cluster)` in `pollOnce(timeout)` in `KafkaConsumer`, right?

I agree with you that there will be unnecessary expected IO. However, according to the API description of `pollOnce(timeout)`, it is expected to do one round of polling. Simply making the change you suggested would violate this API.

It is possible that we spend more time on thinking whether it is beneficial to make this API change. But this requires thorough thinking which is not a simple job. I also worry that this discussion will become longer and longer. 

Given the problem you just mentioned, how about I revert to my earlier approach, i.e. we discard already fetched data upon OffsetOutOfRange exception and fetch the entire data again. What do you think?